server/Arcadia.Assistant.CSP/Vacations/DatabaseVacationsCache.cs

```diff
@@ -0,0 +1,121 @@
+﻿namespace Arcadia.Assistant.CSP.Vacations
+{
+    using System.Collections.Generic;
+    using System.Linq;
+    using Arcadia.Assistant.Calendar.Abstractions;
+
+    internal class DatabaseVacationsCache
+    {
+        private readonly Dictionary<string, CalendarEventWithApprovals> cache;
+
+        public DatabaseVacationsCache(Dictionary<string, CalendarEventWithApprovals> values)
+        {
+            this.cache = new Dictionary<string, CalendarEventWithApprovals>(values);
+        }
+
+        public CalendarEventWithApprovals this[string index]
+        {
+            get => this.cache[index];
+            set => this.cache[index] = value;
+        }
+
+        public Diff Difference(Dictionary<string, CalendarEventWithApprovals> values)
+        {
+            var createdEvents = new List<CalendarEventWithApprovals>();
+            var updatedEvents = new List<CalendarEventWithApprovals>();
+            var approvalsUpdatedEvents = new List<CalendarEventWithApprovals>();
+
+            foreach (var @event in values.Values)
+            {
+                if (!this.cache.ContainsKey(@event.CalendarEvent.EventId))
+                {
+                    if (@event.CalendarEvent.Status == VacationStatuses.Requested)
+                    {
+                        if (!@event.Approvals.Any())
```

Comment:
can it be BOTH newly created but with some approvals added?

Reply:
If event was "created", it will also flow all the flow needed to get next approval and be added to pending actions.